portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -33,17 +34,31 @@
 public class DataServlet extends HttpServlet {
   private static final String COMMENT_ENTITY = "Comment";
   private static final String COMMENT_CONTENT_PROPERTY = "content";
+  private static final String COMMENT_TIMESTAMP_PROPERTY = "timestamp";
+  private static final String NUM_COMMENTS_PARAMETER = "num-comments";
 
   private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
   private Gson gson = new Gson();
-  private Query commentsQuery = new Query(COMMENT_ENTITY);
+  private Query commentsQuery = new Query(COMMENT_ENTITY).addSort("timestamp", SortDirection.DESCENDING);
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get comments from datastore
+    // get number of comments to display
+    int numComments = getNumCommentsParameter(request);
+    if (numComments == -1) {
+      response.setContentType("text/html");
+      response.getWriter().println("Please enter a non-negative integer.");
+      return;
+    }
+
+    // Get comments from datastore, and add numComments number of comments to comments list
     PreparedQuery results = datastore.prepare(commentsQuery);
     List<String> comments = new ArrayList<>();
+    int numCommentsAdded = 0;
     for (Entity entity : results.asIterable()) {
+      if (numCommentsAdded++ >= numComments) {
```

Comment:
I can see how it looks confusing... my code originally looked like yours, but that loop looks like it will always add at least one entity to the list, even if numCommentsRequested is 0. I'll try and make it more readable, let me know if you would like me to change it some more!